### PR TITLE
Fix for missing disconnect when using gmp version autodetection

### PR DESCRIPTION
**Pull request by BaaaZen**
**What**:
We found out that only half of the SSH connections made were closed after leaving the `Gmp` context. After debugging the code we found out, that there is a problem when using the GMP version autodetection in combination with the use of the context manager.

**Why**:
This patch is a proposal to fix the problem with missing connection closes when using the GMP version autodetection.
The problem is caused by returning a new object (instead of self) from the context manager. Thus the connection is opened by the "new" object (the instance of the detected GMP protocol version) and closed by the detector object (`Gmp` class). So the connection was never closed, because the `Gmp` instance never opened it, and the "new" object never got a disconnect.

Maybe this is correlated with our support ticket `GS-854`?

**How**:
Instead of using the `Gmp` class directly for a context there is a new property `supported_gmp` that should be used. This circumvents the problem.

**Checklist**:
TBD

### CODE REVIEW

File: gvm/protocols/gmp.py [hunk 1]
```diff
@@ -113,9 +113,12 @@ def determine_supported_gmp(self) -> SUPPORTED_GMP_VERSIONS:
...
+    @property
+    def supported_gmp(self):
+        return self.determine_supported_gmp()
 
-        return gmp
+    def __enter__(self):
+        raise GvmClientError(
+            'You may not use Gmp for context directly. Please use property '
+            'supported_gmp instead.'
+        )
```
Comment (y0urself) [inline gvm/protocols/gmp.py:124]:
```suggestion
        )

```

**Review by y0urself — CHANGES_REQUESTED** [review]

**Comment by y0urself** [comment]
We think that this could be solved in a more beautiful way by just defining an `__exit__()` method that closes the connection, if still existing?

**Comment by BaaaZen** [comment]
IMHO that won't work. When opening a context like this:
```
with Gmp(connection=connection) as gmp:
    print(gmp.get_version())
```
the `__enter__` of the `Gmp` class creates a new object that opens the connection.
When leaving the context, the `__exit__` of the `Gmp` class is called, **not** the `__exit__` of the returned new instance (here in the `gmp` variable). So the `disconnect()` is not called for the connected new instance but for the not connected `Gmp` class.

The `__exit__` method has no knowlege of the new instance that the `__enter__` method returned. So the only way to solve this in our opinion is to move the generating of the new instance out of the context.

**Comment by bjoernricks** [comment]
PR #660 addresses the problem without changing the API.

**Pull request closed without merging**